src/nvim/diff.c

```diff
@@ -1733,25 +1743,713 @@ void diff_clear(tabpage_T *tp)
+      initialize_compareline2(dp, b0, p0, b1);
+    } else if (p == DFPATH2_COMPARE01) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].compare[b1]=p1;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].compare[b0]=p0;
+      p1++, p0++;
+      initialize_compareline2(dp, b0, p0, b1);
+      initialize_compareline2(dp, b1, p1, b0);
+    } else if (p == DFPATH2_SKIP1) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].newline = true;
+      p1++;
+      initialize_compareline2(dp, b1, p1, b0);
+    }
+  }
+  for (int i = 0; i < 2; i++) {
+    for (int j = 0; j < (dp->df_count[b1]+1); j++) {
+      xfree(df_pathmatrix2[i][j].df_path2);
+    }
+    xfree(df_pathmatrix2[i]);
+  }
+  xfree(df_pathmatrix2);
+}
+
+/// The 3d case (for 3 buffers) of the algorithm implemented when diffopt
+/// 'linematch' is enabled. The algorithm constructs a 3d tensor to
+/// compare a diff between 3 buffers. The dimmensions of the tensor are
+/// the length of the diff in each buffer plus 1 A path is constructed by
+/// moving from one edge of the cube/3d tensor to the opposite edge.
+/// Motions from one cell of the cube to the next represent decisions. In
+/// a 3d cube, there are a total of 7 decisions that can be made,
+/// represented by the enum path3_choice which is defined in
+/// buffer_defs.h a comparison of buffer 0 and 1 represents a motion
+/// toward the opposite edge of the cube with components along the 0 and
+/// 1 axes.  a comparison of buffer 0, 1, and 2 represents a motion
+/// toward the opposite edge of the cube with components along the 0, 1,
+/// and 2 axes. A skip of buffer 0 represents a motion along only the 0
+/// axis. For each action, a point value is awarded, and the path is
+/// saved for reference later, if it is found to have been the optimal
+/// path. The optimal path has the highest score.  The score is
+/// calculated as the summation of the total characters matching between
+/// all of the lines which were compared. The structure of the algorithm
+/// is that of a dynamic programming problem.  We can calculate a point
+/// i,j,k in the cube as a function of i-1, j-1, and k-1. To find the
+/// score and path at point i,j,k, we must determine which path we want
+/// to use, this is done by looking at the possibilities and choosing
+/// the one which results in the local highest score.  The total highest
+/// scored path is, then in the end represented by the cell in the
+/// opposite corner from the start location.  The entire algorithm
+/// consits of populating the 3d cube with the optimal paths from which
+/// it may have came.  However, we cannot apply the general 3d case
+/// before first populating the edges and the surfaces of the cube.
+/// Therefore, there are several sets of if / else statements inside the
+/// main loops which determine which case to evaluate.
+/// @param dp
+void linematch_3buffers(diff_T * dp)
```

Comment:
It's worth marking all the internal functions as` static`.

Reply:
If they are all made as static, then it won't be possible to use them via FFI. It's probably worth keeping the important ones as non-static.

Reply:
I'm not familiar with FFI (foreign function interface?) or the applications here, can you recommend specifically what should be static and what not?

Reply:
FFI allows lua plugins to call C functions. Currently [gitsigns](https://github.com/lewis6991/gitsigns.nvim) uses it to call `xdl_diff` (which is also called in `diff.c`) to run diffs in-process (as opposed to spawning an external diff process).

It would be desirable if a function could be provided similar to `xdl_diff` that gets the benefits of linewise diffs. That way the signs from gitsigns will align with linewise diffs. Note `diff.c` uses `xdl_diff` with the `outf` callback but gitsigns uses it with the `hunk_func` callback which gets called a lot less since it omits context lines which aren't necessary for gitsigns as it already knows the buffer contents; it just needs to know the line numbers of the generated hunks.

Here's the full [implemenation](https://github.com/lewis6991/gitsigns.nvim/blob/main/lua/gitsigns/diff.lua) in gitsigns that uses FFI to call `xdl_diff`. It wraps it up into `function M.run_diff(fa, fb, diff_algo)` where `fa` and `fb` are the file contents you want to diff and returns an array of all the hunks.

There is also this PR that attempts to provide Lua C bindings for xdiff.  An xdl_diff like function would help with that too.

Reply:
It looks like this could be accomplished with a wrapper function around xdl_diff that calls xdl_diff, collects the necessary output (diff hunk line numbers and count in each buffer, the same information collected by "emit_cb.outf" for constructing the diff hunks in "diff_read" diff.c), then using this information to call "linematch_2buffers".

problems to overcome:

	-- "linematch_2buffers" depends on a diff hunk struct "diff_T" being already constructed before the algorithm runs, and it depends on "ml_get_buf" to retrieve the lines to generate the comparison scores with the levenshtein distance. Ideally, there may be a way to create a temporary diff_T struct, and set of temporary buffers to retrieve this same information without significantly modifying "linematch_2buffers" to achieve this.

	-- "linematch_2buffers" / "linematch_3buffers" algorithm is only called when the diff is scrolled into view (or file is edited). This is done by calling the function inside "diff_check", which is called when drawing the lines on the screen. By only running the algorithm when a diff is scrolled into view or if it is updated, the algorithm runs only when necessary. If there was a huge file with hundreds of diffs, and every time the file is updated, the "linematch_2buffers" would run hundreds of times each time just one part of the file is changed, and this might get laggy. The simple wrapper function idea I suggested would depend on calling the "linematch_2buffers" on every diff hunk, no matter if it is scrolled into view or not, then passing the information on to your lua code, so this may be very laggy.

I'm not sure exactly the difference in output from your lua code that uses the "hunk_func" callback of xdl_diff instead of "outf", used in diff.c. What is the difference exactly? I've noticed in "xdiff_out", the function passed as "outf" for xdl_diff, only the first line (header) is actually used, the rest of the output in the subsequent function calls which contains the text lines of the file and + and - signs is ignored. you say that calling hunk_func omits these context lines? it seems that in that case hunk_func should be used in xdiff_out also, because it ignores the context lines anyway.

Do you have suggestions on how to proceed? Do you recommend trying to solve these problems in the same pull request or a new one?